[SDL-0234] Proxy Library RPC Generation (#1273)
smartdevicelink/sdl_java_suite: utils/generator/README.md

> # Proxy Library RPC Generator

Reviewer: Can we make the readme file more concise similar to how the iOS one?
https://github.com/vladmu/sdl_ios/blob/SDL-0234-Proxy-Library-RPC-Generation/generator/README.md

It is currently very lengthy and it will be hard to keep it updated in the future whenever we make an update to the generator.

Author: @bilal-alsharifi it is lengthy due to numerous full examples with big copyright comments included in the document, will the removing of those examples solve the problem?

Reviewer 2: If we remove all full examples as @vladmu suggested, README size will decrease 3 times, from 2450 lines to only 850 lines which is the same as we have in sdl_ios. @bilal-alsharifi please confirm if this looks good for you.

Reviewer: I think samples are important to explain the rules. iOS also has sample codes. We can use fewer samples, pick classes with fewer properties as samples, and remove the license header form the samples in the readme to make the readme file shorter and easier to maintain. Please use the iOS one as a reference:
https://github.com/vladmu/sdl_ios/tree/SDL-0234-Proxy-Library-RPC-Generation/generator

Author: @bilal-alsharifi that iOS document was based on this Java document, and it just does not include the full examples, almost all other sections and their order are similar. So it is hard to understand what you want to change here. I've proposed above to remove FULL examples from the document, all SHORT samples stay in place. Please review simplified readme @ksologubov provided [here (7da2be7)](https://github.com/smartdevicelink/sdl_java_suite/blob/7da2be78ce3d0b0e6e2381e33d972cabcd9377ec/utils/generator/README.md)  and if that doesn't fit your expectation, please provide us more details about what you would like to change exactly. Thank you!

Reviewer: Thanks, @vladmu for the details. I looked at the android simplified readme. I don't see any generated java sample code at all in it. I see snippets from the templates but not java code. However, in iOS, there are some full but concise samples like [this](https://github.com/vladmu/sdl_ios/tree/SDL-0234-Proxy-Library-RPC-Generation/generator#request-functions-sdlrpcrequest) for example. We just would like the readme to be easy to read but not extremely long.
As I suggested earlier, using fewer samples, picking classes with fewer properties as samples, and removing the license header from the samples may reduce the size of the file but still keep it self contained.

Reviewer: Like for example, before today's updates, the readme had [three](https://github.com/smartdevicelink/sdl_java_suite/tree/744d2bfd76beb8f9adcc3158dca176ef05da270a/utils/generator#below-is-the-full-example-of-the-struct-class-with-simple-and-enum-parameters-inside) lengthy full sample for the `<struct>` section. After the update, they are all removed. I think having one concise sample that covers the basic cases will be a good balance between the two cases.

Reviewer 2: Ok, got it, thank you, will update with basic samples

Reviewer 2: @bilal-alsharifi Hi Bilal, I updated README with samples, could you please check

Reviewer: Thank you. I will review that.

 * POSSIBILITY OF SUCH DAMAGE.
 */
```
Where `[year]` in the copyright line is the current year.

## `<enum>`

The name of the class is the value from the `"name"` attribute of `<enum>`.

The class should have the next JavaDoc comment:
```java
/**
 * [description]
 *
 * @deprecated
 * @since SmartDeviceLink [since_version]
 * @see [see_reference]
 */
```
Where:
* `[description]` is `<description>` of the current `<enum>`, if exists.
* `@deprecated` indicates the deprecation state if the `"deprecated"` attribute exists and is "true".
* `@since` should be present, if the `"since"` attribute exists, and `[since_version]` is the `Major.Minor.Patch` formatted value of this attribute.
* `@see` shows the custom reference in `[see_reference]`.

The class should have the `@Deprecated` decorator if the `"deprecated"` attribute of the `<enum>` exists and is "true".

The set of `<element>` should be mapped to the set of Enum constants. Based on the `<element>` attributes, constants could be with or without fields.

The following list are general rules for constant names and its fields:
1. The `"name"` attribute of `<element>` is the default name of the constant.

>     * if the name starts from digit, the leading `_` (underscore) symbol should be added.

Reviewer: I think that case doesn't exist anymore in the code. Also, I believe, the case where we replace "-" with "_" is not mentioned in the readme.

Reviewer 2: No, this exists because there is Enums which will fail without it, e.g. BitsPerSample
element name="8_BIT" should be converted to _8_BIT

Reviewer: I meant the other point:
> if the name starts with _ (underscore) symbol and the next character is a letter of alphabet, the leading _ (underscore) symbol should be removed.

do we still have this case?

Reviewer: Also, I am not sure why, but the CI doesn't pass on this PR.

Reviewer 2: Thanks, Bilal, updated on both questions, waiting for current build to see the problem.

Reviewer: Awesome. It passes now. Thanks.

Reviewer 2: As I see from history it was failing either on android artifacts download with 504 error or git checkout which is far from RPC generator changes.

Reviewer: Yes, that is strange. it fails again now and keeps failing when I restart it.
Failing on downloading the Android artifacts is normal and happens rarely sometimes but simply restarting the action fixes it. However, I have never seen it failing on checkout before. I am guessing it is something has to do with checking out out the submodule that was added here:
https://github.com/smartdevicelink/sdl_java_suite/blob/07ad3eee3682e0ce9306b523821f0d3e85dbcc95/.github/workflows/android.yml#L13

Maybe try to remove that line (just for testing) and see if it passes the checkout stage. There are other alternatives available online to checkout submodules if that is what is causing the issue.

Reviewer 2: Agree, removed my android.yml changes, let's see if build will be stable without checking out submodules

Reviewer: It is still failing on the checkout stage which is strange. I am still guessing it is something has to do with submodules addition because it passes on all other PRs. Can you please try to remove the `.gitmodules` (just temporarily) to see if that makes it pass?

Reviewer: It could be a bug in the `checkout` action. I have made PR #1305  to update the actions `android.yml` to the latest version. It is merged into `develop` now. Can you merge the latest develop into your branch to see if that fixes the issue?

Reviewer 2: I was able to fix this with checkout@master action. Details is in last comments of https://github.com/actions/checkout/issues/81
In short, checkout@v2 did not work with submodules at all now, but updated submodule processing was already comitted to master of checkout action. It will be moved soon to v2 tag

Reviewer: Awesome. Thanks for doing that research. We can keep the master tag for now and change to v2 later when the fix is merged.

Reviewer: @ksologubov can we try v2 tag again? it seems to be fixed now.

Reviewer 2: @bilal-alsharifi  set to v2, works good

Reviewer: Awesome. Thanks

>     * if the name consists of `-` (dash) then it replaced with `_` (underscore)

Reviewer: ```suggestion
    * if the name contains a `-` (dash) then it should be replaced with `_` (underscore)
```

# Proxy Library RPC Generator

> ## Overview

Reviewer: The `README.md` seems to be a bit outdated now. Can we update it to reflect the latest changes like mapping removal and enum simplifications?

Reviewer 3: @vladmu could you have look

Reviewer 2: Readme updated

This script provides the possibility to auto-generate Java code based on a given SDL MOBILE_API XML specification.

## Requirements

> The script requires Python 3.5 or later pre-installed in the system. This is the minimal Python 3 version that has not reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: @bilal-alsharifi the code is compatible with 3.5 thus minimum required version is 3.5. This part just notifies that the generator will not work with previous versions. Even after 3.5 will reach EoL, the code will be compatible with that version until somebody puts the incompatible code. Do we still need this fix here?

> The script requires Python 3.5 or later pre-installed in the system. This is the minimal Python 3 version that has not reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Reviewer: but stating that we are going to support a relatively old version will prevent us from using any of the features in the newer python version. Maybe 3.8 is very new. I suggest something in the middle like 3.6 or 3.7. Let's wait on doing that though until we discuss it with the other teams to make sure all platforms align on that.

> The script requires Python 3.5 or later pre-installed in the system. This is the minimal Python 3 version that has not reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: @bilal-alsharifi the document states "3.5 or later" therefore, I don't think this note prevents you from using the newest features if needed. This only about the current state of the generator, but if someone puts incompatible features in the future, the document should be adjusted simultaneously.

> The script requires Python 3.5 or later pre-installed in the system. This is the minimal Python 3 version that has not reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: ```suggestion
The script requires **Python 3** pre-installed on the host system. The minimal supported Python 3 version is **3.7.6**. It may work on versions back to 3.5 (the minimal version that has not yet reached [the end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)), but this is not supported and may break in the future.
```

> The script requires Python 3.5 or later pre-installed in the system. This is the minimal Python 3 version that has not reached the end-of-life (https://devguide.python.org/devcycle/#end-of-life-branches).

Author: @bilal-alsharifi done in https://github.com/smartdevicelink/sdl_java_suite/pull/1273/commits/7599c4180f36c420179a492e5eb09cd6c34e5eda the same way as discussed with @joeljfischer in iOS PR https://github.com/smartdevicelink/sdl_ios/pull/1556#discussion_r390300250

 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
```
Where `[year]` in the copyright line is the current year.

## `<enum>`

The name of the class is the value from the `"name"` attribute of `<enum>`.

The class should have the next JavaDoc comment:
```java
/**
 * [description]
 *
 * @deprecated
 * @since SmartDeviceLink [since_version]
 * @see [see_reference]
 */
```
Where:
* `[description]` is `<description>` of the current `<enum>`, if exists.
* `@deprecated` indicates the deprecation state if the `"deprecated"` attribute exists and is "true".

> * `@since` should be present, if the `"since"` attribute exists, and `[since_version]` is the `Major.Minor.Patch` formatted value of this attribute.

Author: @kshala-ford `<element name="FR-CA" internal_name="FR_CA">` will not have the `@since` JavaDoc as described in the document. Should it appear based on the parent `since` attribute?

> * `@since` should be present, if the `"since"` attribute exists, and `[since_version]` is the `Major.Minor.Patch` formatted value of this attribute.

Reviewer 4: Not really because @since is added to the class’s inline documentation but it would better appear in the ide documentary. I’m okay to leave as is.

* `[name]` is the `"name"` attribute of `<param>`.
* `[description]` is `<description>` of the `<param>`, if exists.
* `@deprecated` indicates the deprecation state if the `"deprecated"` attribute exists and is "true".
* `@since` should be present, if the `"since"` attribute exists, and `[since_version]` is the `Major.Minor.Patch` formatted value of this attribute.
* `@see` shows the custom reference in `[see_reference]`, if it's defined in the custom mapping.

The field definition should have the `@Deprecated` decorator if the `"deprecated"` attribute of the `<param>` exists and is "true".

The Struct class contains 3 different constructors:
* without parameters.
* with `Hashtable` parameter.
* with all required parameters, based on `"mandatory"` attribute of the `<param>`

### Constructor without parameters

Template:
```java
    /**
     * Constructs a new [name] object
     */
    public [name]() { }
```
Where `[name]` is the value from the `"name"` attribute of `<struct>`.

### Constructor with `Hashtable` parameter

This constructor requires the import of `Hashtable` class
```java
import java.util.Hashtable;
```

Template:
```java
    /**

>      * Constructs a new [name] object indicated by the Hashtable parameter\

Reviewer 2: fixed

     *
     * @param hash The Hashtable to use
     */
    public [name](Hashtable<String, Object> hash) {
        super(hash);
    }
```
Where `[name]` is the value from the `"name"` attribute of `<struct>`.

> ### Constructor with all required parameters, based on `"mandatory"` attribute of the `<param>`

Reviewer 2: > I am curious about the implementation. What's the order of the constructor parameters? Is it following the order of mandatory params elements in the struct?

Yes, it's following the order of mandatory parameters

1. The getter method:
    * If `"type"` attribute of the `<param>` has the one of `Boolean`, `Integer` or `String`
        * the getter should call and return the corresponding `getBoolean`, `getInteger` or `getString` method, the parameter of that method is the value of the corresponding static field described above;
    * If `"type"` attribute of the `<param>` is `Float`:
        * the getter should call the `getValue` method, the parameter of that method is the value of the corresponding static field described above;
        * the getter should return `SdlDataTypeConverter.objectToFloat(object)` where the `object` is the value previously received from `getValue`;
    * If the `<param>` has the `"type"` attribute value as the one of `<enum>` or `<struct>` name:
        * The getter should call and return the result of the `getObject` method, where the first parameter is the corresponding Struct or Enum class, the second is the value of the corresponding static field described above;

Setter template:
```java
    /**
     * Sets the [name].
     * [description]
     *

>      * @param [name]

Reviewer 2: Hello @bilal-alsharifi 
parameter description moved to getter/setter/constructor javadoc as you described

>      * @param [name]

Reviewer: Thanks @ksologubov for making the change

>      * @param [name]

Reviewer: @ksologubov  would it be possible to make the param description starts on the same line as the param? that applies to the @retrun as well.

>      * @param [name]

Reviewer: also, if we can move the `@since comments` from the `public static final String` keys to the getter/setters as well. Because developers will look at the docs for the methods not the keys

>      * @param [name]

Reviewer: So the expected result for `iconResumed` for example will be something like this :

```
public static final String KEY_ICON_RESUMED = "iconResumed";

/**
 * Sets the iconResumed.
 *
 * @param iconResumed Existence of apps icon at system. If true, apps icon was resumed at system.
 * If false, apps icon is not resumed at system
 *
 * @since SmartDeviceLink 5.0.0
 */
public void setIconResumed(Boolean iconResumed) {
    setParameters(KEY_ICON_RESUMED, iconResumed);
}

/**
 * Gets the iconResumed.
 *
 * @return Boolean Existence of apps icon at system. If true, apps icon was resumed at system.
 * If false, apps icon is not resumed at system
 *
 * @since SmartDeviceLink 5.0.0
 */
public Boolean getIconResumed() {
    return getBoolean(KEY_ICON_RESUMED);
}
```

>      * @param [name]

Reviewer 2: Thanks for example Bilal, I'm going to update this

>      * @param [name]

Reviewer 2: @bilal-alsharifi Generator was corrected per your comments
